### PR TITLE
[ENG-4035] Add DomainReference model

### PR DESCRIPTION
**Pull request by adlius**
## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

### CODE REVIEW

**Comment by Johnetordoff** [comment]
Are we sure we don't want to make this a `ManyToMany` relationship like the contributor models? Like for example:
```
    contributors = models.ManyToManyField(
        settings.AUTH_USER_MODEL,
        through=InstitutionalContributor,
        related_name='institutions'
    )
```
This PR creates the correct SQL tables, not I'm not 100% sure it is the practice best for Django, otherwise it's good.

File: osf/models/notable_domain.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        return super().save(*args, **kwargs)
+
+    def __repr__(self):
+        return f'<{self.__class__.__name__}: {self.domain} ({self.Note(self.note).name})>'
+
+    def __str__(self):
+        return repr(self)
+
+class DomainReference(BaseModel):
+    referent_content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
```
Comment (Johnetordoff) [inline osf/models/notable_domain.py:44]: nit: `referent_content_type` and `referent_object_id` maybe is more readable as `reference` in this case?

**Comment by adlius** [comment]
> Are we sure we don't want to make this a `ManyToMany` relationship like the contributor models? Like for example:
> 
> ```
>     contributors = models.ManyToManyField(
>         settings.AUTH_USER_MODEL,
>         through=InstitutionalContributor,
>         related_name='institutions'
>     )
> ```
> 
> This PR creates the correct SQL tables, not I'm not 100% sure it is the practice best for Django, otherwise it's good.

I believe the reason we do a separate model as a through table is because we can't add M2M relationship to models of different content types. What we did for the contributors is more or less a compromise: there is one user model, it needs to have M2M relationship to many different models, so we add different `through` models for each different M2M relationship. This is also helpful when we need to serialize them through the API.

If we use `GenericForeignKey`, we can avoid specifying a different `through` model for each different M2M relationship. Also we don't need to define the inverse relationship on different models, because I don't think we need that inverse relationship at all.

File: osf/models/notable_domain.py (hunk 1 above)
Comment (jwalz) [reply osf/models/notable_domain.py:44]: `referrer` is where you may want to go in that case

File: osf/migrations/0251_domainreference.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+# Generated by Django 1.11.29 on 2022-09-19 17:11
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django_extensions.db.fields
+import osf.models.base
+
+
+class Migration(migrations.Migration):
```
Comment (jwalz) [inline osf/migrations/0251_domainreference.py:11]: Merge this into the above migration, please

**Comment by jwalz** [comment]
> > Are we sure we don't want to make this a `ManyToMany` relationship like the contributor models? Like for example:
> > ```
> >     contributors = models.ManyToManyField(
> >         settings.AUTH_USER_MODEL,
> >         through=InstitutionalContributor,
> >         related_name='institutions'
> >     )
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > This PR creates the correct SQL tables, not I'm not 100% sure it is the practice best for Django, otherwise it's good.
> 
> I believe the reason we do a separate model as a through table is because we can't add M2M relationship to models of different content types. What we did for the contributors is more or less a compromise: there is one user model, it needs to have M2M relationship to many different models, so we add different `through` models for each different M2M relationship. This is also helpful when we need to serialize them through the API.
> 
> If we use `GenericForeignKey`, we can avoid specifying a different `through` model for each different M2M relationship. Also we don't need to define the inverse relationship on different models, because I don't think we need that inverse relationship at all.

Nailed it. And if we find we do need direct access from, say, the Project, we can add a GenericRelationship there that adds manager functionality that will let us access the DomainReferences a given project is a part of

**Review by jwalz — APPROVED** [review]

**Pull request merged**